src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs

```diff
@@ -0,0 +1,182 @@
+         * ill-formed subsequence can be deduced from these values.
+         *
+         * If 'replaceInvalidSequences' is true, the method is guaranteed never to return
+         * OperationStatus.InvalidData. If 'isFinalBlock' is true, the method is
+         * guaranteed never to return OperationStatus.NeedMoreData.
+         */
+
+        public static OperationStatus FromUtf16(ReadOnlySpan<char> source, Span<byte> destination, out int numCharsRead, out int numBytesWritten, bool replaceInvalidSequences = true, bool isFinalBlock = true)
+        {
+            int originalSourceLength = source.Length;
+            int originalDestinationLength = destination.Length;
+            OperationStatus status = OperationStatus.Done;
+
+            // In a loop, this is going to read and transcode one scalar value at a time
+            // from the source to the destination.
+
+            while (!source.IsEmpty)
+            {
+                status = Rune.DecodeUtf16(source, out Rune firstScalarValue, out int charsConsumed);
+
+                switch (status)
+                {
+                    case OperationStatus.NeedMoreData:
+
+                        // Input buffer ended with a high surrogate. Only treat this as an error
+                        // if the caller told us that we shouldn't expect additional data in a
+                        // future call.
+
+                        if (!isFinalBlock)
+                        {
+                            goto Finish;
+                        }
+
+                        status = OperationStatus.InvalidData;
+                        goto case OperationStatus.InvalidData;
+
+                    case OperationStatus.InvalidData:
+
+                        // Input buffer contained invalid data. If the caller told us not to
+                        // perform U+FFFD replacement, terminate the loop immediately and return
+                        // an error to the caller.
+
+                        if (!replaceInvalidSequences)
+                        {
+                            goto Finish;
+                        }
+
+                        firstScalarValue = Rune.ReplacementChar;
+                        goto default;
+
+                    default:
+
+                        // We know which scalar value we need to transcode to UTF-8.
+                        // Do so now, and only terminate the loop if we ran out of space
+                        // in the destination buffer.
+
+                        if (firstScalarValue.TryEncodeToUtf8Bytes(destination, out int bytesWritten))
+                        {
+                            source = source.Slice(charsConsumed); // don't use Rune.Utf8SequenceLength; we may have performed substitution
+                            destination = destination.Slice(bytesWritten);
+                            status = OperationStatus.Done; // forcibly set success
```

Comment:
Why do we have to set success here?

Reply:
When the loop terminates due to end of buffer it'll return the most recently seen `OperationStatus`. If we successfully performed a replacement, we want this to be `OperationStatus.Done` instead of `OperationStatus.InvalidData`.